metrics/provider.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package metrics
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	metricsNamespace = "mattermod"
+	httpNamespace    = "requests"
+	cronNamespace    = "cron"
+	githubNamespace  = "github"
+
+	defaultPrometheusTimeoutSeconds = 60
+)
+
+// Provider is the interface that exposes the communication with the metrics system
+// this interface should be implemented by the different providers we want to include
+type Provider interface {
+	// ObserverHTTPRequestDuration stores the elapsed time for an HTTP request
+	ObserveHTTPRequestDuration(method, handler, statusCode string, elapsed float64)
+	// IncreaseWebhookRequest increases the counter for the webhook requests
+	// identified by name
+	IncreaseWebhookRequest(name string)
+
+	// ObserveGithubRequestDuration stores the elapsed time for github requests
+	ObserveGithubRequestDuration(method, handler, statusCode string, elapsed float64)
+	// IncreaseGithubCacheHits stores the number of cache hits when a github request
+	// is done. The information is stored using the HTTP method and the request handler
+	IncreaseGithubCacheHits(method, handler string)
+	// IncreaseGithubCacheMisses stores the number of cache misses when a github request
+	// is done. The information is stored using the HTTP method and the request handler
+	IncreaseGithubCacheMisses(method, handler string)
+
+	// ObserverCronTaskDuration stores the elapsed time for a cron task
+	ObserveCronTaskDuration(name string, elapsed float64)
+	// IncreaseCronTaskErrors stores the number of errors for a cron task
+	IncreaseCronTaskErrors(name string)
+}
+
+// PrometheusProvider is the implementation of the Provider interface
+// to send metrics to Prometheus
+type PrometheusProvider struct {
+	Registry *prometheus.Registry
+
+	httpRequestsDuration *prometheus.HistogramVec
+	webhookEvents        *prometheus.CounterVec
+
+	cronTasksDuration *prometheus.HistogramVec
```

Comment:
Is this metric going to be meaningful if we collect for all of the cron jobs. AFAIK we have several cron jobs? I mean we should observe these duration per cron job type.